internal/controller/etcd/reconcile_spec.go

```diff
@@ -0,0 +1,174 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package etcd
+
+import (
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	ctrlutils "github.com/gardener/etcd-druid/internal/controller/utils"
+	"github.com/gardener/etcd-druid/internal/operator"
+	"github.com/gardener/etcd-druid/internal/operator/component"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *Reconciler) triggerReconcileSpecFlow(ctx component.OperatorContext, etcdObjectKey client.ObjectKey) ctrlutils.ReconcileStepResult {
+	reconcileStepFns := []reconcileFn{
+		r.recordReconcileStartOperation,
+		r.syncEtcdResources,
+		r.updateObservedGeneration,
+		r.recordReconcileSuccessOperation,
+		r.removeOperationAnnotation,
+	}
+
+	for _, fn := range reconcileStepFns {
+		if stepResult := fn(ctx, etcdObjectKey); ctrlutils.ShortCircuitReconcileFlow(stepResult) {
+			return r.recordIncompleteReconcileOperation(ctx, etcdObjectKey, stepResult)
+		}
+	}
+	ctx.Logger.Info("Finished spec reconciliation flow")
+	return ctrlutils.ContinueReconcile()
+}
+
+func (r *Reconciler) removeOperationAnnotation(ctx component.OperatorContext, etcdObjKey client.ObjectKey) ctrlutils.ReconcileStepResult {
+	etcd := &druidv1alpha1.Etcd{}
+	if result := r.getLatestEtcd(ctx, etcdObjKey, etcd); ctrlutils.ShortCircuitReconcileFlow(result) {
+		return result
+	}
+	if _, ok := etcd.Annotations[v1beta1constants.GardenerOperation]; ok {
+		ctx.Logger.Info("Removing operation annotation")
+		withOpAnnotation := etcd.DeepCopy()
```

Comment:
> but I was wondering like how we have get for partialObjectMeta, do we a patch as well for the same.

Thanks for the suggestion. @unmarshall this is something we should try out.

> Also additionally we are calling this right after the update to the ETCD spec, can't we just reuse that instead of fetching it again as we are in the same reconcile loop still looping through reconcile steps?

It's always a good practice to fetch the latest possible version of the etcd resource before trying to make any changes on it. `removeOperationAnnotation` has no knowledge of the operations performed before or after it, and rightly so. So this function should independently fetch the etcd (or partialObjectMetadata) before performing any operations on it.

Reply:
>It's always a good practice to fetch the latest possible version of the etcd resource before trying to make any changes on it. removeOperationAnnotation has no knowledge of the operations performed before or after it, and rightly so. So this function should independently fetch the etcd (or partialObjectMetadata) before performing any operations on it.

Agree. 👍🏽